src/Grpc.NetCore.HttpClient/Internal/GrpcCallScope.cs

```diff
@@ -0,0 +1,72 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Grpc.Core;
+
+namespace Grpc.NetCore.HttpClient.Internal
+{
+    internal class GrpcCallScope : IReadOnlyList<KeyValuePair<string, object>>
```

Comment:
All this work to avoid an allocation of an IDictionary 😄

Reply:
Plus the cached ToString for log writers.

Reply:
Who is the logging expert these days? I'd like to double check I am using log scope correctly.

Reply:
Hmmm @brennanconroy?

Reply:
This class looks good, but some of the usage is questionable.

For example:
https://github.com/grpc/grpc-dotnet/pull/236/files#diff-9ca9d0e546493f5532bd42e33e273dc1R389
You dispose the scope right after this, so any logs that happen after the method goes async wont have the scope attached to them.

Reply:
> You dispose the scope right after this, so any logs that happen after the method goes async wont have the scope attached to them.

They do have the scope. This message that the response headers have been received happens after the call goes async and I see the scope in the console output:

```
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[2]
      => GrpcMethodType:ServerStreaming,GrpcUri:/grpc.testing.TestService/StreamingOutputCall
      Response headers received.
```

I'm not really sure how a scope works. I initially called logger.AddScope when the call started, cached the disposable response on GrpcCall, and then disposed it when the call ended, but that didn't work.

Reply:
I'm going to guess it is an async local and async/await handles flowing the execution context.